### ISSUE TITLE
OONI Probe Android is performing duplicate requests to the API server for checking the existence of measurements and adding unnecessary load to our infrastructure

### ISSUE THREAD

**Issue by hellais** [ooni/probe#1034]
While investigating why the OONI API was going down very frequently today, I was looking at the logs and noticed that the logic of OONI Probe Android for retrieving measurements is not behaving as nice as it could.

Namely we see that:

1. When it sees 500 series error it is not backing off, but rather triggering a follow up swarm of requests
2. Even when it does finally receive a 200 OK response, it is duplicating the requests.

Below is an excerpt of the log:

```
[CLIENT_IP_1] - - [07/Mar/2020:14:12:46 +0000] "GET /api/v1/measurements?report_id=20200222T191142Z_AS3292_S9AU3vqDF8jyNB3jRZHs2EHLHwHLKN3gLv7CcHcL7VRKA4oHdp HTTP/2.0" 200 948 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191142Z_AS3292_S9AU3vqDF8jyNB3jRZHs2EHLHwHLKN3gLv7CcHcL7VRKA4oHdp HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191142Z_AS3292_S9AU3vqDF8jyNB3jRZHs2EHLHwHLKN3gLv7CcHcL7VRKA4oHdp HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191142Z_AS3292_S9AU3vqDF8jyNB3jRZHs2EHLHwHLKN3gLv7CcHcL7VRKA4oHdp HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191142Z_AS3292_S9AU3vqDF8jyNB3jRZHs2EHLHwHLKN3gLv7CcHcL7VRKA4oHdp HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191142Z_AS3292_S9AU3vqDF8jyNB3jRZHs2EHLHwHLKN3gLv7CcHcL7VRKA4oHdp HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191142Z_AS3292_S9AU3vqDF8jyNB3jRZHs2EHLHwHLKN3gLv7CcHcL7VRKA4oHdp HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191142Z_AS3292_S9AU3vqDF8jyNB3jRZHs2EHLHwHLKN3gLv7CcHcL7VRKA4oHdp HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191142Z_AS3292_S9AU3vqDF8jyNB3jRZHs2EHLHwHLKN3gLv7CcHcL7VRKA4oHdp HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191142Z_AS3292_S9AU3vqDF8jyNB3jRZHs2EHLHwHLKN3gLv7CcHcL7VRKA4oHdp HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 200 1872 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:47 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 200 1870 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 503 190 "-" "okhttp/4.0.1" "api.ooni.io"
[CLIENT_IP_1] - - [07/Mar/2020:14:12:48 +0000] "GET /api/v1/measurements?report_id=20200222T191147Z_AS3292_Hc7cJ2lq9VFMaxUhmLCeLe7TIBmdWXTbEQouPwNVdMF8TqDNjl HTTP/2.0" 200 1871 "-" "okhttp/4.0.1" "api.ooni.io"
```

In order to fix this you should:

* [ ] Implement some sort of backoff logic when a request fails (ex. you see a 503 error, wait first 2 seconds before retrying, then 4 seconds, then 8 seconds, then 16 and then just give up)
* [ ] Avoid duplicating the requests if you see a 200 series message. This is a bug.

**Comment by lorenzoPrimi** [comment]
The function `measurements?` is called in two points:
Here https://github.com/ooni/probe-android/blob/master/app/src/main/java/org/openobservatory/ooniprobe/model/database/Measurement.java#L209
on every app start only if 86400 has passed from the last call
https://github.com/ooni/probe-android/blob/master/app/src/main/java/org/openobservatory/ooniprobe/common/Application.java#L45

And here
https://github.com/ooni/probe-android/blob/master/app/src/main/java/org/openobservatory/ooniprobe/activity/MeasurementDetailActivity.java#L176
When an user opens the `MeasurementDetailActivity` and the relative measurement has a report file on disk.

Lastly when the user "open json file" and the measurement it not on disk it will be downloaded from the internet
https://github.com/ooni/probe-android/blob/master/app/src/main/java/org/openobservatory/ooniprobe/activity/TextActivity.java#L126

I don't see any problem with the first one, the second one can be called multiple times if the relative measurement is marked not uploaded and this can be fixed, but I still don't see how the requests can be multiplied apart if the user enter and exit the screen on his own.
In every call onError does nothing, I never recall the function.

Unless there is a problem with how OKHTTP handles response 500.

**Comment by lorenzoPrimi** [comment]
Another difference with iOS is that in Android we never pass the `input` item
https://github.com/ooni/probe-ios/blob/master/ooniprobe/Model/Database/Measurement.mm#L112
Can this lead to multiple calls in case of a web connectivity test with many urls?

**Comment by lorenzoPrimi** [comment]
No, because we only care about element 0 
https://github.com/ooni/probe-android/blob/master/app/src/main/java/org/openobservatory/ooniprobe/client/callback/GetMeasurementsCallback.java#L18

**Comment by lorenzoPrimi** [comment]
I think the problem can be in the `selectMeasurementsWithJson` function in case of web connectivity 
https://github.com/ooni/probe-android/blob/master/app/src/main/java/org/openobservatory/ooniprobe/model/database/Measurement.java#L106

as we query
```
	return SQLite.select().from(Measurement.class)
		.where(Measurement_Table.is_failed.eq(false))
		.and(Measurement_Table.is_rerun.eq(false))
		.and(Measurement_Table.is_done.eq(true))
		.and(OperatorGroup.clause()
				.or(Measurement_Table.is_uploaded.eq(true))
				.or(Measurement_Table.report_id.isNotNull())
		);
```

And for a test like websites this function will be called for every input, not for every test.
This was added in 2.2.0 not in 2.3.0

**Comment by lorenzoPrimi** [comment]
So basically we need to 
- Add input to the API call (Android) to make the function equal to iOS, in this case the parameter input is always sent.

AND

- Call this function once per report_id and not per measurement in case of web connectivity (both platforms)
And somehow manage to delete all the jsons associated to that report_id as websites saves a json locally for every input.

I'm still not convinced about the second part, it can require a lot of code refactoring.

**Comment by lorenzoPrimi** [comment]
Another issue found, the call to the backend to delete json when they are published is tone every 86400 ms not s.

**Comment by hellais** [comment]
Ok so let's take a step back a minute. I will write down some assumptions I think are true and please correct me if they are wrong.

**Why is this call being done?**

My understanding is that the purpose of the `https://api.ooni.io/api/v1/measurements?report_id=XXX&input=YYY` call is to understand if a certain measurements (or set of measurements) has been uploaded to the collector.

If it has been uploaded to the collector, then this means you are safe to delete the measurements on disk.

Assuming this is the case, I believe you don't need to specify the `input` at all. If you see that a given **report_id** is present in the API server, you should consider **all the measurements** in that result set to have been uploaded and processed.

From what you have written above, I derive that you are doing one call for every single measurements. This scales very badly. Imagine you have just run the full test list (in this release we have added support for this option, so it should be supported method) which contains 1000-2000 URLs.
I suspect what the app currently would do is perform 1000-2000 HTTP requests to api.ooni.io to determine if each of these measurements has been uploaded, while it could simply do a single request for the `report_id` and then consider all the measurements uploaded.

Moreover, from looking at the logs the status code returned from the API is ignored and actually in the even of an error a retry is triggered immediately. This creates a problem somewhat similar to a [thundering herd problem](https://en.wikipedia.org/wiki/Thundering_herd_problem), where once the API goes down and starts returning an error each client keeps hitting it with a retry.

Rather than doing this, the app should implement an exponential backoff in doing its retries.

So to recap, in order to fix this problem, assuming my assumption above are correct (if they are not please do let me know), I believe we should:

1. Add an exponential backoff when handling API side-errors
2. Perform only 1 lookup per report_id (`input` is irrelevant, we don't care about it)

These can either be done in a single PR/patch or as two separate ones depending on the complexity of implementation.
The exponential backoff is probably the most important thing to ship ASAP.

**Comment by hellais** [comment]
On top of these two, I would also add another third element, which is that the OONI Probe mobile apps should identify themselves with a clearly distinguishable user agent.

I suggest using the following:

```
ooniprobe-android/X.Y.Z
ooniprobe-ios/X.Y.Z
```

**Comment by lorenzoPrimi** [comment]
> Assuming this is the case, I believe you don't need to specify the `input` at all. If you see that a given **report_id** is present in the API server, you should consider **all the measurements** in that result set to have been uploaded and processed.

Not all measurement in a result belongs to the same "set". One measurement can be failed and re ran so it will have a different report_id.

I  also thought about doing it once per report_id but this will change a lot the logic in the app (now it's a for loop, then will be a more complex algo) and I don't think can be done as "hotfix" but will take a bit more of time.

**Comment by lorenzoPrimi** [comment]
> On top of these two, I would also add another third element, which is that the OONI Probe mobile apps should identify themselves with a clearly distinguishable user agent.
> 
> I suggest using the following:
> 
> ```
> ooniprobe-android/X.Y.Z
> ooniprobe-ios/X.Y.Z
> ```

in the GET request? like client=ooniprobe-ios-2.3.0 ?

**Comment by lorenzoPrimi** [comment]
> Moreover, from looking at the logs the status code returned from the API is ignored and actually in the even of an error a retry is triggered immediately. This creates a problem somewhat similar to a [thundering herd problem](https://en.wikipedia.org/wiki/Thundering_herd_problem), where once the API goes down and starts returning an error each client keeps hitting it with a retry.

I don't see any retry at all in the app code. When a request fails it goes to the next.

**Comment by lorenzoPrimi** [comment]
So basically I should edit the function `selectMeasurementsWithJson` also to group them (measurements with json) by report_id and delete all the jsons relative to that report_id, am I correct?

**Comment by bassosimone** [comment]
> I don't see any retry at all in the app code. When a request fails it goes to the next.

Are you sure, @lorenzoPrimi, that okhttp doesn't retry for us?

**Comment by lorenzoPrimi** [comment]
This can be possible, I'm gonna research how the library behaves

**Comment by lorenzoPrimi** [comment]
Great input, we have to disable this.
https://medium.com/inloopx/okhttp-is-quietly-retrying-requests-is-your-api-ready-19489ef35ace

So @hellais if we want an hotfix delivered today we can
- Disable retrying
- Fix the ms to s bug so it will retry once per day 
https://github.com/ooni/probe-android/commit/ca3a018d96e8a76b5026422ca019d1b163f07396

And straight after that we do an algorithm that query once per report_id
